spec/linter-eslint-spec.js

```diff
@@ -84,23 +84,23 @@ async function getNotification(expectedMessage) {
 }
 
 async function makeFixes(textEditor) {
-  return new Promise(async (resolve) => {
+  const editorReloaded = new Promise((resolve) => {
     // Subscribe to the file reload event
-    const editorReloadSub = textEditor.getBuffer().onDidReload(async () => {
-      editorReloadSub.dispose()
-      // File has been reloaded in Atom, notification checking will happen
-      // async either way, but should already be finished at this point
+    const reloadSubscription = textEditor.getBuffer().onDidReload(() => {
+      reloadSubscription.dispose()
       resolve()
     })
+  })
 
-    // Now that all the required subscriptions are active, send off a fix request
-    atom.commands.dispatch(atom.views.getView(textEditor), 'linter-eslint:fix-file')
-    const expectedMessage = 'Linter-ESLint: Fix complete.'
-    const notification = await getNotification(expectedMessage)
+  // Now that subscription is active, send off a fix request
```

Comment:
How about this?
```js
// Editor reload subscription is now active, send ...
```

Reply:
Yeah, that's fine for me. I only even messed with that comment to make it singular. The old comment falsely implied multiple subscriptions, while there is only one subscription per call to `makeFixes`.

Reply:
There are actually two, `getNotification` creates a subscription on new Notifications in order to grab the one we are waiting for. _Technically_ there is a race condition between dispatching the command and the Notification subscription being added, but the few _us_ that takes is far outweighed by the time it takes to call out to ESLint to make the fixes.

Reply:
Hmmm. I didn't catch that.

The race between `getNotifications` and `textEditor.getBuffer().onDidReload` proved quite enough to make tests flaky. Based on the comments, I assume the previous developer believed that it was "technically" a race condition.

So I guess we need to 
```
subscribe to reload
subscribe to notifications
dispatch commands
await notifications
return reload promise
```
?

Reply:
> The race between `getNotifications` and `textEditor.getBuffer().onDidReload` proved quite enough to make tests flaky.

Do you have any basis for this? Although the changes in this PR are more correct, I haven't seen any issues with this in a while.

The only falkiness I know of was actually due to `onDidReload` being completely broken in the default Jasmine environment. (["previous" developer](https://github.com/AtomLinter/linter-eslint/commit/a9269577da08c4de93a341146221b6c1dbcf5ce9#diff-15b3e4c0d78ceab6572aceecc5b4bd6cR204) 👋)

This is fixed in `jasmine-fix` [here](https://github.com/steelbrain/jasmine-fix/blob/c7bc55cc9d6de1f4493f3f7a1930c60b80c00557/index.js#L15-L23).

----

I couldn't think of a good way to have a nice `async` function to get a specific notification that was usable in multiple spots due to the callback nature of the `onDid____` functions.

Reply:
> The only falkiness I know of was actually due to onDidReload being completely broken in the default Jasmine environment. 

Since "flakiness" is defined by it's inconsistent nature, there's no real way to know whether it was fixed by that patch. But since it's related to promises and in particular `onDidReload`... maybe? 

> Do you have any basis for this? Although the changes in this PR are more correct, I haven't seen any issues with this in a while.

This fix was based on which specs [reported failure](https://github.com/AtomLinter/linter-eslint/pull/1009#issuecomment-329997867) in randomly failing CircleCI builds. This was about a week before your jasmine fix, and `makeFix` does call `onDidReload`. So honestly I have no idea now.

Maybe I'm just being a perfectionist? I don't see why you would want to even consider allowing race conditions in specs. Are the _potentially_ saved microseconds during testing really worth having indeterminate tests?

It's all running in a single thread, so everything is at best concurrent, and never parallel. If you actually even manage to save some time (which you certainly might fail to accomplish, due to the nature v8's parsing and compiling optimizations), the potential savings is quite negligible.

Reply:
> This fix was based on which specs [reported failure](https://github.com/AtomLinter/linter-eslint/pull/1009#issuecomment-329997867) in randomly failing CircleCI builds. This was about a week before your jasmine fix, and `makeFix` does call `onDidReload`. So honestly I have no idea now.

Ah, if I remember correctly that was actually what triggered my deep investigation into what in the world was going on with the specs and why they were so flaky 😉.

> Maybe I'm just being a perfectionist? I don't see why you would want to even consider allowing race conditions in specs.

Ideally we wouldn't at all, that's not being a perfectionist, just writing good code.

> Are the _potentially_ saved microseconds during testing really worth having indeterminate tests?

I'm not sure what this has to do with performance?

The current design of the `getNotifications` function is only to get the code that is used in many places all into a single spot. The issue with doing it correctly is that each place this function is currently being called would have to be changed to setup the observer and wrap the callback itself. If you have an alternative design for this I'd love to hear it! I don't like relying on this race condition either, no matter how impossible it _should_ be to trigger it.

Reply:
Proof that current master branch still has flaky tests.

![image](https://user-images.githubusercontent.com/844400/34884063-8fcd982c-f789-11e7-97fb-526f4cda27f8.png)